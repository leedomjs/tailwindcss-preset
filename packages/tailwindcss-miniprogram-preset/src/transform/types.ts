/**
 * fraction array
 * @example ['1/2', '1/3']
 */
export type Fraction = Array<string>

/**
 * tailwindcss params
 * @example
 * width: {
 *    '1-dv-2': '50%'
 * }
 *
 */
export interface CommonParam {
  [key: string]: string
}

/**
 * tailwindcss extend params
 * @example
 * extend: {
 *    width: {
 *       '1-dv-2': '50%'
 *    }
 * }
 */
export interface PercentExtend {
  [key: string]: CommonParam
}

/**
 * tailwindcss fontSize params
 * @example
 * lineHeight: {
 *    'sm': {
 *      font-size: 0.875rem;
 *      line-height: 1.25rem;
 *     }
 * }
 */
export interface FontSizeValue {
  [key: string]: [string, { lineHeight: string }]
}
