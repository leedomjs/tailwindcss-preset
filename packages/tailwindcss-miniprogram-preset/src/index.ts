import type { Config } from 'tailwindcss'
import {
  extend,
  fontSize,
  lineHeight,
  spacing,
} from './transform'

import type { MiniProgramConfig } from './transform/types'

export default (opt: MiniProgramConfig = {
  enable: true,
}): Config => opt.enable ? {
  content: [],
  theme: {
    /**
     * refer to https://tailwindcss.com/docs/customizing-spacing
     *
     * By default the spacing scale is inherited by the padding, margin, width, minWidth, maxWidth, height, minHeight, maxHeight, gap, inset, space, translate, scrollMargin, and scrollPadding core plugins.
     *
     * tailwindcss 0.25rem = 4px
     * IPhone6 1px = 2rpx
     * So with tailwindcss in mp, 0.25rem = 8rpx
     */
    spacing,
    fontSize,
    lineHeight,
    screens: [], // mp doesn't need screens.
    extend: {
      ...extend,
    },
  },
  corePlugins: {
    preflight: false, // mp doesn't need base styles.
  },
} : {
  content: [],
}

export type * from './transform/types'
