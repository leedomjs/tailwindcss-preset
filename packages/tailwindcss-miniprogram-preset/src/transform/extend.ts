import type { CommonParam, Fraction, PercentExtend } from './types'

function getAllFractions(number: number): Fraction {
  const fractions: Fraction = []
  for (let i = 1; i < number; i++) {
    fractions.push(`${i}/${number}`)
  }
  return fractions
}

function generatePercent() {
  const percent = [2, 3, 4, 5, 6, 12]
  return ['width', 'height', 'size'].reduce((obj: PercentExtend, item) => {
    let classesList = {}
    for (const val of percent) {
      const classes = getAllFractions(val).reduce((map: CommonParam, val) => {
        const [a, b] = val.split('/')
        map[val.replace('/', '-dv-')] = `${((+a / +b) * 100).toFixed(Number.isInteger((+a / +b) * 100) ? 0 : 6)}%`
        return map
      }, {})
      classesList = { ...classesList, ...classes }
      obj[item] = classesList
    }
    return obj
  }, {})
}

const extend = generatePercent()

export default extend
