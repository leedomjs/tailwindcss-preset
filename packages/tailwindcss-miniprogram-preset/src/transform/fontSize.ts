import { fontSize } from 'tailwindcss/defaultTheme'
import type { FontSizeValue } from './types'

export default Object.entries(fontSize).reduce((obj: FontSizeValue, [key, value]) => {
  const [size, { lineHeight }] = value
  const rpxSize = `${(+size.replace('rem', '')) * 32}rpx`
  const rpxLineHeight: string = lineHeight === '1' ? lineHeight : `${(+lineHeight.replace('rem', '')) * 32}rpx`
  obj[key] = [rpxSize, { lineHeight: rpxLineHeight }]
  return obj
}, {})
