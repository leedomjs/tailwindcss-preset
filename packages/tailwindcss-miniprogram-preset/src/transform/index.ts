export { default as spacing } from './spacing'
export { default as fontSize } from './fontSize'
export { default as lineHeight } from './lineHeight'
export { default as extend } from './extend'
export type * from './types'
