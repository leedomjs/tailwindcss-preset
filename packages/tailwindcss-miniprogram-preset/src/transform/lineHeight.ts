import { lineHeight } from 'tailwindcss/defaultTheme'
import type { CommonParam } from './types'

export default Object.entries(lineHeight).reduce((obj: CommonParam, [key, value]) => {
  if (Number.isInteger(+key)) {
    obj[key] = `${+value.replace('rem', '') * 32}rpx`
  } else {
    obj[key] = value
  }
  return obj
}, {})
