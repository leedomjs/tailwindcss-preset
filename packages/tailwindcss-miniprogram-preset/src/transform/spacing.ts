import { spacing } from 'tailwindcss/defaultTheme'
import type { CommonParam } from './types'

export default Object.entries(spacing).reduce((obj: CommonParam, [key, value]) => {
  if (!['0', 'px', '0.5', '1.5', '2.5', '3.5'].includes(key)) {
    obj[key] = `${+value.replace('rem', '') * 32}rpx`
  } else if (['0', 'px'].includes(key)) {
    obj[key] = value
  } else {
    obj[key.replace('.', '-')] = `${+value.replace('rem', '') * 32}rpx`
  }
  return obj
}, {})
