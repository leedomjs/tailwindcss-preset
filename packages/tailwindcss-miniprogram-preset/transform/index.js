const spacing = require('./spacing')
const fontSize = require('./fontSize')
const lineHeight = require('./lineHeight')
const extend = require('./extend')

module.exports = {
  spacing,
  fontSize,
  lineHeight,
  extend,
}
