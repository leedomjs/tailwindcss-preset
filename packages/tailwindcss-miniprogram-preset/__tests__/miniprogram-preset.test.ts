import { describe, expect, it, test } from 'vitest'
import { extend, fontSize, lineHeight, spacing } from '../src/transform'
import mpPreset from '../src/index'

describe('miniprogram-preset', () => {
  test('extend', () => {
    expect(extend).toMatchSnapshot()
  })

  test('spacing', () => {
    expect(spacing).toMatchSnapshot()
  })

  test('fontSize', () => {
    expect(fontSize).toMatchSnapshot()
  })

  test('lineHeight', () => {
    expect(lineHeight).toMatchSnapshot()
  })

  test('mpPreset', () => {
    expect(mpPreset({ enable: true })).toMatchSnapshot()
  })

  it('should be no-config', () => {
    expect(mpPreset({ enable: false })).toEqual({
      content: [],
    })
  })
})
