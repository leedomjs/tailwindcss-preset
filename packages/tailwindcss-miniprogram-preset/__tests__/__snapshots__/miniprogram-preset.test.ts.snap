// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`miniprogram-preset > extend 1`] = `
{
  "height": {
    "1-dv-12": "8.333333%",
    "1-dv-2": "50%",
    "1-dv-3": "33.333333%",
    "1-dv-4": "25%",
    "1-dv-5": "20%",
    "1-dv-6": "16.666667%",
    "10-dv-12": "83.333333%",
    "11-dv-12": "91.666667%",
    "2-dv-12": "16.666667%",
    "2-dv-3": "66.666667%",
    "2-dv-4": "50%",
    "2-dv-5": "40%",
    "2-dv-6": "33.333333%",
    "3-dv-12": "25%",
    "3-dv-4": "75%",
    "3-dv-5": "60%",
    "3-dv-6": "50%",
    "4-dv-12": "33.333333%",
    "4-dv-5": "80%",
    "4-dv-6": "66.666667%",
    "5-dv-12": "41.666667%",
    "5-dv-6": "83.333333%",
    "6-dv-12": "50%",
    "7-dv-12": "58.333333%",
    "8-dv-12": "66.666667%",
    "9-dv-12": "75%",
  },
  "size": {
    "1-dv-12": "8.333333%",
    "1-dv-2": "50%",
    "1-dv-3": "33.333333%",
    "1-dv-4": "25%",
    "1-dv-5": "20%",
    "1-dv-6": "16.666667%",
    "10-dv-12": "83.333333%",
    "11-dv-12": "91.666667%",
    "2-dv-12": "16.666667%",
    "2-dv-3": "66.666667%",
    "2-dv-4": "50%",
    "2-dv-5": "40%",
    "2-dv-6": "33.333333%",
    "3-dv-12": "25%",
    "3-dv-4": "75%",
    "3-dv-5": "60%",
    "3-dv-6": "50%",
    "4-dv-12": "33.333333%",
    "4-dv-5": "80%",
    "4-dv-6": "66.666667%",
    "5-dv-12": "41.666667%",
    "5-dv-6": "83.333333%",
    "6-dv-12": "50%",
    "7-dv-12": "58.333333%",
    "8-dv-12": "66.666667%",
    "9-dv-12": "75%",
  },
  "width": {
    "1-dv-12": "8.333333%",
    "1-dv-2": "50%",
    "1-dv-3": "33.333333%",
    "1-dv-4": "25%",
    "1-dv-5": "20%",
    "1-dv-6": "16.666667%",
    "10-dv-12": "83.333333%",
    "11-dv-12": "91.666667%",
    "2-dv-12": "16.666667%",
    "2-dv-3": "66.666667%",
    "2-dv-4": "50%",
    "2-dv-5": "40%",
    "2-dv-6": "33.333333%",
    "3-dv-12": "25%",
    "3-dv-4": "75%",
    "3-dv-5": "60%",
    "3-dv-6": "50%",
    "4-dv-12": "33.333333%",
    "4-dv-5": "80%",
    "4-dv-6": "66.666667%",
    "5-dv-12": "41.666667%",
    "5-dv-6": "83.333333%",
    "6-dv-12": "50%",
    "7-dv-12": "58.333333%",
    "8-dv-12": "66.666667%",
    "9-dv-12": "75%",
  },
}
`;

exports[`miniprogram-preset > fontSize 1`] = `
{
  "2xl": [
    "48rpx",
    {
      "lineHeight": "64rpx",
    },
  ],
  "3xl": [
    "60rpx",
    {
      "lineHeight": "72rpx",
    },
  ],
  "4xl": [
    "72rpx",
    {
      "lineHeight": "80rpx",
    },
  ],
  "5xl": [
    "96rpx",
    {
      "lineHeight": "1",
    },
  ],
  "6xl": [
    "120rpx",
    {
      "lineHeight": "1",
    },
  ],
  "7xl": [
    "144rpx",
    {
      "lineHeight": "1",
    },
  ],
  "8xl": [
    "192rpx",
    {
      "lineHeight": "1",
    },
  ],
  "9xl": [
    "256rpx",
    {
      "lineHeight": "1",
    },
  ],
  "base": [
    "32rpx",
    {
      "lineHeight": "48rpx",
    },
  ],
  "lg": [
    "36rpx",
    {
      "lineHeight": "56rpx",
    },
  ],
  "sm": [
    "28rpx",
    {
      "lineHeight": "40rpx",
    },
  ],
  "xl": [
    "40rpx",
    {
      "lineHeight": "56rpx",
    },
  ],
  "xs": [
    "24rpx",
    {
      "lineHeight": "32rpx",
    },
  ],
}
`;

exports[`miniprogram-preset > lineHeight 1`] = `
{
  "10": "80rpx",
  "3": "24rpx",
  "4": "32rpx",
  "5": "40rpx",
  "6": "48rpx",
  "7": "56rpx",
  "8": "64rpx",
  "9": "72rpx",
  "loose": "2",
  "none": "1",
  "normal": "1.5",
  "relaxed": "1.625",
  "snug": "1.375",
  "tight": "1.25",
}
`;

exports[`miniprogram-preset > mpPreset 1`] = `
{
  "content": [],
  "corePlugins": {
    "preflight": false,
  },
  "theme": {
    "extend": {
      "height": {
        "1-dv-12": "8.333333%",
        "1-dv-2": "50%",
        "1-dv-3": "33.333333%",
        "1-dv-4": "25%",
        "1-dv-5": "20%",
        "1-dv-6": "16.666667%",
        "10-dv-12": "83.333333%",
        "11-dv-12": "91.666667%",
        "2-dv-12": "16.666667%",
        "2-dv-3": "66.666667%",
        "2-dv-4": "50%",
        "2-dv-5": "40%",
        "2-dv-6": "33.333333%",
        "3-dv-12": "25%",
        "3-dv-4": "75%",
        "3-dv-5": "60%",
        "3-dv-6": "50%",
        "4-dv-12": "33.333333%",
        "4-dv-5": "80%",
        "4-dv-6": "66.666667%",
        "5-dv-12": "41.666667%",
        "5-dv-6": "83.333333%",
        "6-dv-12": "50%",
        "7-dv-12": "58.333333%",
        "8-dv-12": "66.666667%",
        "9-dv-12": "75%",
      },
      "size": {
        "1-dv-12": "8.333333%",
        "1-dv-2": "50%",
        "1-dv-3": "33.333333%",
        "1-dv-4": "25%",
        "1-dv-5": "20%",
        "1-dv-6": "16.666667%",
        "10-dv-12": "83.333333%",
        "11-dv-12": "91.666667%",
        "2-dv-12": "16.666667%",
        "2-dv-3": "66.666667%",
        "2-dv-4": "50%",
        "2-dv-5": "40%",
        "2-dv-6": "33.333333%",
        "3-dv-12": "25%",
        "3-dv-4": "75%",
        "3-dv-5": "60%",
        "3-dv-6": "50%",
        "4-dv-12": "33.333333%",
        "4-dv-5": "80%",
        "4-dv-6": "66.666667%",
        "5-dv-12": "41.666667%",
        "5-dv-6": "83.333333%",
        "6-dv-12": "50%",
        "7-dv-12": "58.333333%",
        "8-dv-12": "66.666667%",
        "9-dv-12": "75%",
      },
      "width": {
        "1-dv-12": "8.333333%",
        "1-dv-2": "50%",
        "1-dv-3": "33.333333%",
        "1-dv-4": "25%",
        "1-dv-5": "20%",
        "1-dv-6": "16.666667%",
        "10-dv-12": "83.333333%",
        "11-dv-12": "91.666667%",
        "2-dv-12": "16.666667%",
        "2-dv-3": "66.666667%",
        "2-dv-4": "50%",
        "2-dv-5": "40%",
        "2-dv-6": "33.333333%",
        "3-dv-12": "25%",
        "3-dv-4": "75%",
        "3-dv-5": "60%",
        "3-dv-6": "50%",
        "4-dv-12": "33.333333%",
        "4-dv-5": "80%",
        "4-dv-6": "66.666667%",
        "5-dv-12": "41.666667%",
        "5-dv-6": "83.333333%",
        "6-dv-12": "50%",
        "7-dv-12": "58.333333%",
        "8-dv-12": "66.666667%",
        "9-dv-12": "75%",
      },
    },
    "fontSize": {
      "2xl": [
        "48rpx",
        {
          "lineHeight": "64rpx",
        },
      ],
      "3xl": [
        "60rpx",
        {
          "lineHeight": "72rpx",
        },
      ],
      "4xl": [
        "72rpx",
        {
          "lineHeight": "80rpx",
        },
      ],
      "5xl": [
        "96rpx",
        {
          "lineHeight": "1",
        },
      ],
      "6xl": [
        "120rpx",
        {
          "lineHeight": "1",
        },
      ],
      "7xl": [
        "144rpx",
        {
          "lineHeight": "1",
        },
      ],
      "8xl": [
        "192rpx",
        {
          "lineHeight": "1",
        },
      ],
      "9xl": [
        "256rpx",
        {
          "lineHeight": "1",
        },
      ],
      "base": [
        "32rpx",
        {
          "lineHeight": "48rpx",
        },
      ],
      "lg": [
        "36rpx",
        {
          "lineHeight": "56rpx",
        },
      ],
      "sm": [
        "28rpx",
        {
          "lineHeight": "40rpx",
        },
      ],
      "xl": [
        "40rpx",
        {
          "lineHeight": "56rpx",
        },
      ],
      "xs": [
        "24rpx",
        {
          "lineHeight": "32rpx",
        },
      ],
    },
    "lineHeight": {
      "10": "80rpx",
      "3": "24rpx",
      "4": "32rpx",
      "5": "40rpx",
      "6": "48rpx",
      "7": "56rpx",
      "8": "64rpx",
      "9": "72rpx",
      "loose": "2",
      "none": "1",
      "normal": "1.5",
      "relaxed": "1.625",
      "snug": "1.375",
      "tight": "1.25",
    },
    "screens": [],
    "spacing": {
      "0": "0px",
      "0-5": "4rpx",
      "1": "8rpx",
      "1-5": "12rpx",
      "10": "80rpx",
      "11": "88rpx",
      "12": "96rpx",
      "14": "112rpx",
      "16": "128rpx",
      "2": "16rpx",
      "2-5": "20rpx",
      "20": "160rpx",
      "24": "192rpx",
      "28": "224rpx",
      "3": "24rpx",
      "3-5": "28rpx",
      "32": "256rpx",
      "36": "288rpx",
      "4": "32rpx",
      "40": "320rpx",
      "44": "352rpx",
      "48": "384rpx",
      "5": "40rpx",
      "52": "416rpx",
      "56": "448rpx",
      "6": "48rpx",
      "60": "480rpx",
      "64": "512rpx",
      "7": "56rpx",
      "72": "576rpx",
      "8": "64rpx",
      "80": "640rpx",
      "9": "72rpx",
      "96": "768rpx",
      "px": "1px",
    },
  },
}
`;

exports[`miniprogram-preset > spacing 1`] = `
{
  "0": "0px",
  "0-5": "4rpx",
  "1": "8rpx",
  "1-5": "12rpx",
  "10": "80rpx",
  "11": "88rpx",
  "12": "96rpx",
  "14": "112rpx",
  "16": "128rpx",
  "2": "16rpx",
  "2-5": "20rpx",
  "20": "160rpx",
  "24": "192rpx",
  "28": "224rpx",
  "3": "24rpx",
  "3-5": "28rpx",
  "32": "256rpx",
  "36": "288rpx",
  "4": "32rpx",
  "40": "320rpx",
  "44": "352rpx",
  "48": "384rpx",
  "5": "40rpx",
  "52": "416rpx",
  "56": "448rpx",
  "6": "48rpx",
  "60": "480rpx",
  "64": "512rpx",
  "7": "56rpx",
  "72": "576rpx",
  "8": "64rpx",
  "80": "640rpx",
  "9": "72rpx",
  "96": "768rpx",
  "px": "1px",
}
`;
